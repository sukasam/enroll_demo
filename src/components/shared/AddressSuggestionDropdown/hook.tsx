import { useEffect } from "react";

function useOnClickOutside(
    ref: React.RefObject<HTMLElement>,
    handler: () => void
): void {
    useEffect(() => {
        const listener = (event: MouseEvent | TouchEvent): void => {
            if (!ref.current || ref.current.contains(event.target as Node)) {
                return;
            }
            handler();
        };

        document.addEventListener("mousedown", listener);
        document.addEventListener("touchstart", listener);

        return (): void => {
            document.removeEventListener("mousedown", listener);
            document.removeEventListener("touchstart", listener);
        };
    }, [ref, handler]);
}

export default useOnClickOutside;
