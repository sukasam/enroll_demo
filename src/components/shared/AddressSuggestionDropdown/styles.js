import { css } from "@emotion/react";

export default css`
    font-family: "Inter", Arial, Helvetica, sans-serif;

    .autocomplete {
        .MuiAutocomplete-popupIndicatorOpen {
            transform: rotate(0deg) !important;
        }

        .MuiAutocomplete-root {
            width: 100% !important;
        }

        .MuiAutocomplete-popper {
            top: 16px;
        }
    }

    .dropdown {
        background: #ffffff;
        border: 1px solid #f1f5f8;
        box-shadow: 0px 4px 18px rgba(0, 0, 0, 0.05);
        border-radius: 8px;
        overflow: hidden;
        z-index: 1000;

        .option {
            height: 50px;
            padding: 0 16px;
            display: flex;
            align-items: center;
            cursor: pointer;
            font-size: 12px;
            color: #153862;

            &:hover {
                background: rgba(153, 188, 223, 0.2);
            }

            &:not(:last-child) {
                border-bottom: 1px solid #f1f5f8;
            }
        }
    }

    hr {
        border: none;
        height: 1px;
        display: flex;
        width: 100%;
        background: #003764;
        opacity: 0.1;
        margin: 24px 0;
    }

    .addressSuggestion {
        display: flex;
        justify-content: space-between;
        align-items: center;
    }

    .more {
        padding-left: 10px;
        font-weight: bold;
        display: flex;
        align-items: center;
    }
    .addressSuggestion {
        width: 100%;
    }
`;
