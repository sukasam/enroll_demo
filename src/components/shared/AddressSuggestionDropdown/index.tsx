/** @jsxImportSource @emotion/react */
import { ChevronRight } from "@mui/icons-material";
import { Popper } from "@mui/material";
import { Alpha2, Alpha3 } from "Constants/countryConfig/enums.js";
import { AddressSuggestion, getAddressById } from "Services/addressSuggestions";
import { toAlpha2 } from "Services/locale";
import { useCallback, useEffect, useRef } from "react";
import { useFormContext } from "react-hook-form";
import useOnClickOutside from "./hook";
import styles from "./styles.js";

type AddressSuggestionDropdownProps = {
    suggestedAddresses: Array<AddressSuggestion>;
    selectedCountry: string;
    setSuggestedAddresses: (addresses: Array<AddressSuggestion>) => void;
    setHintsOpen: (open: boolean) => void;
    hintsOpen: boolean;
    anchorEl: HTMLElement | null;
    setLastAddressDebounce: (address: string) => void;
    onSuggestionSelect?: () => void;
};
export default function AddressSuggestionDropdown({
    suggestedAddresses,
    selectedCountry,
    setSuggestedAddresses,
    setHintsOpen,
    hintsOpen,
    anchorEl,
    setLastAddressDebounce,
    onSuggestionSelect
}: AddressSuggestionDropdownProps): React.ReactElement {
    const popperRef = useRef(null);

    const closeHints = useCallback((): void => {
        setSuggestedAddresses([]);
        setHintsOpen(false);
    }, [setSuggestedAddresses, setHintsOpen]);

    useOnClickOutside(popperRef, () => closeHints());

    const { setValue, watch } = useFormContext();
    const country = watch("country");

    useEffect(() => {
        closeHints();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [country]);

    const fillForm = (suggestedAddress: AddressSuggestion): void => {
        const fieldsToUpdate = [
            { name: "address1", value: suggestedAddress.address1 },
            {
                name: "address2",
                value:
                    country === "JP"
                        ? suggestedAddress.city
                        : suggestedAddress.address2
            },
            { name: "city", value: suggestedAddress.city },
            { name: "state", value: suggestedAddress.state },
            { name: "zip", value: suggestedAddress.zip },
            {
                name: "country",
                value: toAlpha2(suggestedAddress.country as Alpha3)
            }
        ];
        fieldsToUpdate.forEach(({ name, value }) => {
            setValue(name, value, { shouldTouch: true, shouldValidate: true });
        });
    };

    const handleSelectAddress = async (
        event: React.MouseEvent | React.KeyboardEvent,
        address: Record<string, any>
    ): Promise<void> => {
        if (
            event.type === "keydown" &&
            (event as React.KeyboardEvent).key !== "Enter" &&
            (event as React.KeyboardEvent).key !== " "
        ) {
            return;
        }

        if (
            address.entries &&
            address.entries >= 1 &&
            address.addressId !== undefined
        ) {
            const newSuggestion = await getAddressById(
                address.addressId,
                (selectedCountry as Alpha2) || "US"
            );
            if (newSuggestion) {
                setSuggestedAddresses(newSuggestion);
                return;
            }
        }
        fillForm(address as AddressSuggestion);
        setLastAddressDebounce(address.address1);
        closeHints();
        onSuggestionSelect?.();
    };

    return (
        <Popper
            ref={popperRef}
            anchorEl={anchorEl}
            css={styles}
            id="address-hints"
            open={hintsOpen}
            placement="bottom-start"
            popperOptions={{
                modifiers: [
                    {
                        name: "offset",
                        options: {
                            offset: [0, 0]
                        }
                    }
                ]
            }}
            style={{ zIndex: 9 }}
        >
            <div className="dropdown">
                {suggestedAddresses.map((address: Record<string, any>) => (
                    <div
                        className="option"
                        key={`${address.address1}-${address.city}-${address.state}-${address.zip}`}
                        role="button"
                        tabIndex={0}
                        onClick={(event): Promise<void> =>
                            handleSelectAddress(event, address)
                        }
                        onKeyDown={(event): Promise<void> =>
                            handleSelectAddress(event, address)
                        }
                    >
                        {address.addressId ? (
                            <div className="addressSuggestion">
                                <div className="address">
                                    {address.address1}
                                </div>
                                <div className="more">
                                    More <ChevronRight />
                                </div>
                            </div>
                        ) : (
                            `${address.address1} ${address.city} ${address.state} ${address.zip}`
                        )}
                    </div>
                ))}
            </div>
        </Popper>
    );
}
