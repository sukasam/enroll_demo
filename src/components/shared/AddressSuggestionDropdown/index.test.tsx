import { render } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import AddressSuggestionDropdown from "./index";

// Mock the dependencies
jest.mock("Services/addressSuggestions", () => ({
    getAddressById: jest.fn()
}));

jest.mock("Services/locale", () => ({
    toAlpha2: jest.fn()
}));

jest.mock("./hook", () => jest.fn());

function MockFormProvider({
    children
}: {
    children: React.ReactNode;
}): JSX.Element {
    const methods = useForm();
    return <FormProvider {...methods}>{children}</FormProvider>;
}

describe("AddressSuggestionDropdown", () => {
    const mockProps = {
        suggestedAddresses: [
            {
                address1: "123 Main St",
                address2: "",
                city: "Anytown",
                state: "ST",
                zip: "12345",
                country: "US"
            },
            {
                address1: "456 Elm St",
                address2: "",
                city: "Othertown",
                state: "OT",
                zip: "67890",
                country: "US"
            }
        ],
        selectedCountry: "US",
        setSuggestedAddresses: jest.fn(),
        setHintsOpen: jest.fn(),
        hintsOpen: true,
        anchorEl: null,
        setLastAddressDebounce: jest.fn()
    };

    it("renders correctly", () => {
        const { asFragment } = render(
            <MockFormProvider>
                <AddressSuggestionDropdown {...mockProps} />
            </MockFormProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly when closed", () => {
        const { asFragment } = render(
            <MockFormProvider>
                <AddressSuggestionDropdown {...mockProps} hintsOpen={false} />
            </MockFormProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with address IDs", () => {
        const propsWithAddressIds = {
            ...mockProps,
            suggestedAddresses: [
                {
                    addressId: "1",
                    address1: "123 Main St",
                    address2: "",
                    city: "Anytown",
                    state: "ST",
                    zip: "12345",
                    country: "US"
                },
                {
                    addressId: "2",
                    address1: "456 Elm St",
                    address2: "",
                    city: "Othertown",
                    state: "OT",
                    zip: "67890",
                    country: "US"
                }
            ]
        };
        const { asFragment } = render(
            <MockFormProvider>
                <AddressSuggestionDropdown {...propsWithAddressIds} />
            </MockFormProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
