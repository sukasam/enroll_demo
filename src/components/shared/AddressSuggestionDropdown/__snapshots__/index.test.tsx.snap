// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressSuggestionDropdown renders correctly 1`] = `<DocumentFragment />`;

exports[`AddressSuggestionDropdown renders correctly when closed 1`] = `<DocumentFragment />`;

exports[`AddressSuggestionDropdown renders correctly with address IDs 1`] = `<DocumentFragment />`;
